test/c/test_sabre_layout.c

```diff
@@ -0,0 +1,325 @@
+    QkTargetEntry *cx_entry = qk_target_entry_new(QkGate_CX);
+    for (uint32_t i = 0; i < num_qubits - 1; i++) {
+        uint32_t qargs[2] = {i, i + 1};
+        double inst_error = 0.0090393 * (num_qubits - i);
+        double inst_duration = 0.020039;
+        qk_target_entry_add_property(cx_entry, qargs, 2, inst_duration, inst_error);
+    }
+
+    qk_target_add_instruction(target, cx_entry);
+
+    QkCircuit *qc = qk_circuit_new(5, 0);
+    for (uint32_t i = 0; i < qk_circuit_num_qubits(qc) - 1; i++) {
+        uint32_t qargs[2] = {0, i + 1};
+        qk_circuit_gate(qc, QkGate_CX, qargs, NULL);
+    }
+    QkTranspileLayout *layout_result =
+        qk_transpiler_pass_standalone_sabre_layout(qc, target, 4, 20, 20, 2025);
+
+    QkOpCounts op_counts = qk_circuit_count_ops(qc);
+    if (op_counts.len != 2) {
+        printf("More than 2 types of gates in circuit, circuit's instructions are:\n");
+        print_circuit(qc);
+        result = EqualityError;
+        goto layout_cleanup;
+    }
+    for (int i = 0; i < op_counts.len; i++) {
+        int swap_gate = strcmp(op_counts.data[i].name, "swap");
+        int cx_gate = strcmp(op_counts.data[i].name, "cx");
+        if (cx_gate != 0 && swap_gate != 0) {
+            printf("Gate type of %s found in the circuit which isn't expected");
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+        if (swap_gate == 0 && op_counts.data[i].count != 2) {
+            printf("Unexpected number of swaps %d found in the circuit.");
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+    uint32_t expected_initial_layout[5] = {1, 0, 2, 3, 4};
+    uint32_t result_initial_layout[5];
+    qk_transpile_layout_initial_layout(layout_result, false, result_initial_layout);
+    for (uint32_t i = 0; i < 5; i++) {
+        if (result_initial_layout[i] != expected_initial_layout[i]) {
+            printf("Initial layout maps qubit %d to %d, expected %d instead", i,
+                   result_initial_layout[i], expected_initial_layout[i]);
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+
+    uint32_t expected_final_layout[5] = {3, 0, 1, 2, 4};
+    uint32_t result_final_layout[5];
+    qk_transpile_layout_output_permutation(layout_result, result_final_layout);
+    for (uint32_t i = 0; i < 5; i++) {
+        if (result_final_layout[i] != expected_final_layout[i]) {
+            printf("Final layout maps qubit %d to %d, expected %d instead", i,
+                   result_final_layout[i], expected_final_layout[i]);
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
+    QkCircuit *expected_circuit = qk_circuit_new(5, 0);
+    uint32_t qargs[2] = {1, 0};
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    qargs[0] = 1;
+    qargs[1] = 2;
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
```

Comment:
Arrays being non-assignable kind of sucks.  I wonder if
```c
uint32_t qargs[2] = {1, 0};
qk_circuit_gate(expected, QkGate_CX, qargs, NULL);
uint32_t qargs[2] = {1, 2};
qk_circuit_gate(expected, QkGate_CX, qargs, NULL);
// ...
```
would be a nicer pattern when it comes to writing docs?  

Though now I mention it, perhaps the fact that we pass a pointer into the stack to `qk_circuit_gate` will cause a C compiler to think that the memory for each `qargs` needs to stay alive until the end of the function, and we'll end up with a humongous stack frame?

Either way, I think we need to work out a neater way to write this sort of thing for documentation in the future, or potentially to provide specialised 2q standard-gate methods, or _something_ to ease the API.

Reply:
Heh, every time I context switch to C I do the same thing, first try reassigning the array like:
```c
qargs = {1, 2}
```
then I try the redefinition like you're suggesting. But it's a compiler error for me locally with "redefinition of ‘qargs’". Then I write it like this. I'm not sure of a better syntax here, but my C is fairly rusty.

We can look at expanding the specialized methods, I think that's really the only other option here really.